File: src/com/suushiemaniac/cubing/bld/util/BruteForceUtil.java
package com.suushiemaniac.cubing.bld.util;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;
import com.suushiemaniac.cubing.alglib.lang.CubicAlgorithmReader;
import com.suushiemaniac.cubing.alglib.lang.NotationReader;
import com.suushiemaniac.cubing.bld.analyze.BldPuzzle;
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType;
import com.suushiemaniac.cubing.bld.model.enumeration.puzzle.CubicPuzzle;

import java.util.*;
import java.util.stream.Collectors;

public class BruteForceUtil {
    public static String[] ALPHABET = {"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};

    public static String[] genBlockString(String[] alphabet, int length, boolean inclusive, boolean mayRepeat) {
        ArrayList<String> moveList = new ArrayList<>();
        if (length < 1) return new String[0];
        else if (length == 1) return alphabet;
        else {
            if (inclusive)
                for (int i = 1; i < length; i++)
                    Collections.addAll(moveList, genBlockString(alphabet, i, false, mayRepeat));
            for (String genBlockMove : genBlockString(alphabet, length - 1, false, mayRepeat))
                for (String blockMove : alphabet)
                    if (mayRepeat || !genBlockMove.contains(blockMove))
                        moveList.add(genBlockMove + blockMove);
        }
        return moveList.toArray(new String[moveList.size()]);
    }

    public static String[] genBlockString(String[] alphabet, int length, boolean mayRepeat) {
        return genBlockString(alphabet, length, false, mayRepeat);
    }

	public static <T> List<List<T>> permute(Collection<T> alphabet, int length, boolean inclusive, boolean mayRepeat) {
		ArrayList<List<T>> moveList = new ArrayList<>();

		if (length < 1) {
			return new ArrayList<>();
		} else if (length == 1) {
			return alphabet.stream()
					.map(Collections::singletonList)
					.map(ArrayList::new)
					.collect(Collectors.toList());
		} else {
			if (inclusive) {
				for (int i = 1; i < length; i++) {
					moveList.addAll(permute(alphabet, i, false, mayRepeat));
				}
			}

			for (List<T> prevPermute : permute(alphabet, length - 1, false, mayRepeat)) {
				for (T blockObj : alphabet) {
					if (mayRepeat || !prevPermute.contains(blockObj)) {
						List<T> nextPermute = new ArrayList<>(prevPermute);
						nextPermute.add(blockObj);

						moveList.add(nextPermute);
					}
				}
			}
		}

		return moveList;
	}

    public static void bruteForceAlg(String lpCase, PieceType type, String[] alphabet, int prune) {
    	int len = 1;

		NotationReader reader = new CubicAlgorithmReader();
		BldPuzzle analyze = CubicPuzzle.THREE_BLD.getAnalyzingPuzzle();

    	while (len < prune) {
			System.out.println("Trying length " + len + "…");
			String[] moves = genBlockString(alphabet, len, true);

    		for (String alg: moves) {
				Algorithm current = reader.parse(alg);

				if (analyze.solves(type, current, lpCase, false)) {
					System.out.println(current.toFormatString());
				}
			}

			len++;
		}
	}

	public static void bruteForceAlg(String lpCase, PieceType type, String[] alphabet) {
    	bruteForceAlg(lpCase, type, alphabet, 21);
	}
}

File: src/main/kotlin/com/suushiemaniac/cubing/bld/util/BruteForceUtil.kt
package com.suushiemaniac.cubing.bld.util

import com.suushiemaniac.cubing.alglib.alg.Algorithm
import com.suushiemaniac.cubing.alglib.alg.SimpleAlg
import com.suushiemaniac.cubing.alglib.move.Move
import com.suushiemaniac.cubing.bld.gsolve.GPuzzle
import com.suushiemaniac.cubing.bld.model.cycle.PieceCycle
import com.suushiemaniac.cubing.bld.model.PieceType

object BruteForceUtil {
    fun <T> Iterable<T>.permute(length: Int, inclusive: Boolean = false, mayRepeat: Boolean = false): Sequence<List<T>> {
        when {
            length < 1 -> return emptySequence()
            length == 1 -> return this.asSequence().map { listOf(it) }
            else -> return generateSequence {
                if (inclusive)
                    for (i in 1 until length)
                        for (subPerm in permute(i, false, mayRepeat))
                            return@generateSequence subPerm

                for (prevPermute in this.permute(length - 1, false, mayRepeat)) {
                    for (blockObj in this) {
                        if (mayRepeat || !prevPermute.contains(blockObj)) {
                            val nextPermute = prevPermute.toMutableList()
                            nextPermute.add(blockObj)

                            return@generateSequence nextPermute // FIXME
                        }
                    }
                }

                return@generateSequence null
            }
        }
    }

    fun <T> Array<T>.permute(length: Int, inclusive: Boolean = false, mayRepeat: Boolean = false): Sequence<List<T>> {
        return this.toList().permute(length, inclusive, mayRepeat)
    }

    fun bruteForceAlg(analyze: GPuzzle, cycle: PieceCycle, type: PieceType, alphabet: Array<Move>, prune: Int = 21): List<Algorithm> {
        val accumulator = mutableListOf<Algorithm>()

        for (len in 1 until prune) {
            val moves = alphabet.permute(len, inclusive = false, mayRepeat = true).map { SimpleAlg(it) }

            for (alg in moves) {
                if (analyze.solves(type, alg, cycle, false)) {
                    accumulator.add(alg)
                }
            }
        }

        return accumulator
    }
}